crates/bevy_winit/src/system.rs

```diff
@@ -618,6 +691,95 @@ pub(crate) fn changed_cursor_options(
     });
 }
 
+/// Identify all windows that had their [`WindowIcon`] handle change and mark them for updating the winit icon
+#[cfg(feature = "custom_window_icon")]
+pub(crate) fn changed_window_icon(
+    mut commands: bevy_ecs::system::Commands,
+    changed_windows: Query<Entity, Changed<WindowIcon>>,
+) {
+    for entity in &changed_windows {
+        commands.entity(entity).insert(WindowIconRefreshNeeded);
+    }
+}
+
+/// Identify all windows that had their [`WindowIcon`] underlying asset change and mark them for updating the winit icon
+#[cfg(feature = "custom_window_icon")]
+pub(crate) fn changed_window_icon_asset(
+    mut commands: bevy_ecs::system::Commands,
+    windows: Query<(Entity, &WindowIcon)>,
+    mut asset_messages: MessageReader<AssetEvent<Image>>,
+) {
+    for event in asset_messages.read() {
+        if let AssetEvent::LoadedWithDependencies { id } | AssetEvent::Modified { id } = event {
+            for (entity, window_icon) in &windows {
+                if window_icon.handle.id() == *id {
+                    commands.entity(entity).insert(WindowIconRefreshNeeded);
+                }
+            }
+        }
+    }
+}
```

Comment:
Currently using naïve loop to identify the windows using the handle corresponding to the changed asset.
Could get fancy with observers to maintain a lookup table or something, but for now I went with the low complexity solution.